server/bin/test_data_importer.py

```diff
@@ -0,0 +1,706 @@
+    return "%s-%s" % (prefix, suffix)
+
+
+
+class Candlepin:
+    # Server configuration
+    host = 'localhost'
+    port = 8443
+    username = 'admin'
+    password = 'admin'
+
+    status_cache = None
+    hosted_test = None
+
+
+    def __init__(self, host, port, username, password):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.password = password
+
+        self.determine_mode()
+
+    def build_url(self, endpoint):
+        # Remove leading slash if it exists
+        if endpoint and endpoint[0] == '/':
+            endpoint = endpoint[1:]
+
+        return "https://%s:%d/candlepin/%s" % (self.host, self.port, endpoint)
+
+    def get(self, endpoint, params=None, headers=None, content_type='text/plain', accept_type='application/json'):
+        return self.request('GET', endpoint, None, params, headers, content_type, accept_type)
+
+    def post(self, endpoint, data=None, params=None, headers=None, content_type='application/json', accept_type='application/json'):
+        return self.request('POST', endpoint, data, params, headers, content_type, accept_type)
+
+    def put(self, endpoint, data=None, params=None, headers=None, content_type='application/json', accept_type='application/json'):
+        return self.request('PUT', endpoint, data, params, headers, content_type, accept_type)
+
+    def delete(self, endpoint, data=None, params=None, headers=None, content_type='application/json', accept_type='application/json'):
+        return self.request('DELETE', endpoint, data, params, headers, content_type, accept_type)
+
+    def request(self, req_type, endpoint, data=None, params=None, headers=None, content_type='application/json', accept_type='application/json'):
+        if params is None: params = {}
+        if headers is None: headers = {}
+
+        # Add our content-type and accept type if they weren't explicitly provided in the headers already
+        if not ci_in(headers, 'Content-Type'): headers['Content-Type'] = content_type
+        if not ci_in(headers, 'Accept'): headers['Accept'] = accept_type
+
+        log.log(LOGLVL_TRACE, "Sending request: %s %s (params: %s, headers: %s, data: %s)" % (req_type, endpoint, params, headers, data))
+
+        response = requests.request(req_type, self.build_url(endpoint),
+            json=data,
+            auth=(self.username, self.password),
+            params=params,
+            headers=headers,
+            verify=False)
+
+        response.raise_for_status()
+        return response
+
+    def status(self):
+        if self.status_cache is None:
+            self.status_cache = self.get('status').json()
+
+        return self.status_cache
+
+    def determine_mode(self):
+        if self.hosted_test is None:
+            self.hosted_test = not self.status()['standalone']
+
+            if self.hosted_test:
+                try:
+                    self.get('hostedtest/alive', accept_type='text/plain')
```

Comment:
BTW: I don't see `hostedtest` in https://www.candlepinproject.org/swagger/?url=candlepin/swagger-3.1.16.json#/ What is it? Is it used in production?

Reply:
The hostedtest stuff is not deployed in the war and can only be added from the unshipped deploy script by specifying -H. It is a set of APIs that populate the data store behind some adapters we use for testing a simulated upstream to represent how hosted operates -- very similar in behavior and design to the cloud registration testing.

The previous test data injector was kind of doing this as well, but was hiding the hosted/standalone separation behind the deprecated ```create_pool_and_subscription```, which tried to do both at the same time to varying degrees of success.

Reply:
Good to know. Thanks. :+1: